src/toil/wdl/wdltoil.py

```diff
@@ -404,51 +703,101 @@ def map_over_files_in_binding(binding: WDL.Env.Binding[WDL.Value.Base], transfor
+        return WDL.Value.Array(value.type.item_type, [map_over_typed_files_in_value(v, transform) for v in value.value], value.expr)
     elif isinstance(value, WDL.Value.Map):
         # This is a map, so recurse on the members of the items, which are tuples (but not wrapped as WDL Pair objects)
         # TODO: Can we avoid a cast in a comprehension if we get MyPy to know that each pair is always a 2-element tuple?
-        return WDL.Value.Map(value.type.item_type, [cast(Tuple[WDL.Value.Base, WDL.Value.Base], tuple((map_over_files_in_value(v, transform) for v in pair))) for pair in value.value], value.expr)
+        return WDL.Value.Map(value.type.item_type, [cast(Tuple[WDL.Value.Base, WDL.Value.Base], tuple((map_over_typed_files_in_value(v, transform) for v in pair))) for pair in value.value], value.expr)
     elif isinstance(value, WDL.Value.Pair):
         # This is a pair, so recurse on the left and right items
-        return WDL.Value.Pair(value.type.left_type, value.type.right_type, cast(Tuple[WDL.Value.Base, WDL.Value.Base], tuple((map_over_files_in_value(v, transform) for v in value.value))), value.expr)
+        return WDL.Value.Pair(value.type.left_type, value.type.right_type, cast(Tuple[WDL.Value.Base, WDL.Value.Base], tuple((map_over_typed_files_in_value(v, transform) for v in value.value))), value.expr)
     elif isinstance(value, WDL.Value.Struct):
         # This is a struct, so recurse on the values in the backing dict
-        return WDL.Value.Struct(cast(Union[WDL.Type.StructInstance, WDL.Type.Object], value.type), {k: map_over_files_in_value(v, transform) for k, v in value.value.items()}, value.expr)
+        return WDL.Value.Struct(cast(Union[WDL.Type.StructInstance, WDL.Type.Object], value.type), {k: map_over_typed_files_in_value(v, transform) for k, v in value.value.items()}, value.expr)
     else:
         # All other kinds of value can be passed through unmodified.
         return value
 
-class WDLInputJob(Job):
+class WDLBaseJob(Job):
+    """
+    Base job class for all WDL-related jobs.
+    """
+
+    def __init__(self, **kwargs: Any) -> None:
+        """
+        Make a WDL-related job.
+
+        Makes sure the global recursive call limit is high enough to allow
+        MiniWDL's extremely deep WDL structures to be pickled. We handle this
+        in the constructor because it needs to happen in the leader and the
+        worker before a job body containing MiniWDL structures can be saved.
+        """
+        super().__init__(**kwargs)
+
+        # The jobs can't pickle under the default Python recursion limit of
+        # 1000 because MiniWDL data structures are very deep.
+        # TODO: Dynamically determine how high this needs to be to serialize the structures we actually have.
+        # TODO: Make sure C-level stack size is also big enough for this.
+        sys.setrecursionlimit(10000)
```

Comment:
How big are the MiniWDL data structures? Would this be enough for larger workflows?

Also, should we lower this limit after we pickle the jobs?

Reply:
I talked to Mike a bit about this; the MiniWDL data structures end up taking the whole parsed WDL with them because they all refer back to the code they came from.

I have no idea whether this will be enough for larger workflows and I'm not sure how to tell other than by trying it. I tried to use some tools to get a look at the structure of the pickled data to try and get an idea of why exactly it needs more than 1000 recursions and I couldn't get the tools working for me.

I'm not sure how we'd detect that the process has pickled the last MiniWDL value it will need to pickle either. We can maybe hook on an object to tell that it is being pickled, but I don't think we can tell that another similar object isn't about to be pickled. And we can't hook the MiniWDL objects themselves, just the Toil jobs that contain them.

We could try something with `__getstate__` and like represent the job as a string of the result of pickling its fields with the limit raised? But that might be fiddly because each job subclass can define new fields that we'd want to bring along.

I want to look into whether we can just ship the WDLs as text and avoid pickling the parsed source code, so I think we are going to revise this piece in the future.